### PR TITLE
Add Cooperative Locking user guide

### PR DESCRIPTION
**Pull request by medb**

### CODE REVIEW

File: gcs/COOPERATIVE_LOCKING.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
+### Introducing GCS Connector Cooperative Locking
+
+We are excited to release Cooperative Locking support for GCS Connector. With
+this release you will be able to isolate and recover directory modification
+operations performed through hadoop fs command line interface on Cloud Storage.
+
+The key motivation behind this feature is a fact that directory mutations are
+not atomic on GCS, which means that competing directory mutations are
+susceptible to race conditions that can lead to data integrity issues (loss,
+duplication, etc).
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: Briefly describe the how cooperative locking works would be helpful.

File: gcs/COOPERATIVE_LOCKING.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: These benefits are hard to understand in this context, I literally don't know what "write-only isolation", "recovery of failed directory operations" mean. Why do I want to isolation, why directory operations could fail and I need to care.

File: gcs/COOPERATIVE_LOCKING.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
+### Introducing GCS Connector Cooperative Locking
+
+We are excited to release Cooperative Locking support for GCS Connector. With
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: Can we add version number?

File: gcs/COOPERATIVE_LOCKING.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+</configuration>
+```
+
+or specify it in directly `hadoop fs` command:
+
+```shell
+hadoop fs -Dfs.gs.cooperative.locking.enable=true ...
+```
+
+### Using Cooperative Locking
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: What I use the feature in Spark/Hadoop, if nothing needs to be done specially, still mention it explicitly.

File: gcs/COOPERATIVE_LOCKING.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+To perform isolated directory modification operations inside the same bucket you
+should use `hadoop fs` command with enabled Cooperative Locking:
+
+```shell
+hadoop fs -{mv|rm} gs://bucket/path/to/dir1 [gs://bucket/path/to/dir2]
+```
+
+### Using FSCK tool
+
+To recover failed directory modification operations performed with enabled
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: When could it fail?

File: gcs/COOPERATIVE_LOCKING.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+```shell
+hadoop --loglevel debug fs -Dfs.gs.cooperative.locking.enable=true -{mv|rm} \
+  gs://bucket/path/to/dir1 [gs://gs://bucket/path/to/dir2]
+hadoop --loglevel debug jar /usr/lib/hadoop/lib/gcs-connector.jar \
+  com.google.cloud.hadoop.fs.gcs.CoopLockFsck \
+  --{check|rollBack|rollForward} gs://<bucket_name> [all|<operation-id>]
+```
+
+Also you may want to inspect `*.log` and `*.lock` files for specific directory
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: Briefly introduce what it looks like in .lock and .log.

File: gcs/COOPERATIVE_LOCKING.md (hunk 4 above)
Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: In "Limitations" section it's mentioned that Cooperative Locking intended for use by human operators.

File: gcs/COOPERATIVE_LOCKING.md (hunk 5 above)
Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: Done.

File: gcs/COOPERATIVE_LOCKING.md (hunk 6 above)
Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: Done.

File: gcs/COOPERATIVE_LOCKING.md (hunk 2 above)
Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: Done.

File: gcs/COOPERATIVE_LOCKING.md (hunk 1 above)
Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: Done.

File: gcs/COOPERATIVE_LOCKING.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    }
+    ```
+
+*   Operation `*.lock` file
+
+    ```shell
+    gsutil cat gs://example-bucket/_lock/20190820T163409.446Z_RENAME_2f6e9ca4-0406-4612-82d5-7f07de81aeb0.lock | jq
+    ```
+
+    Example output:
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: Briefly introduce the meaning of the sample output.

File: gcs/COOPERATIVE_LOCKING.md [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    }
+    ```
+
+*   Operation `*.log` file
+
+    ```shell
+    gsutil cat gs://example-bucket/_lock/20190820T163409.446Z_RENAME_2f6e9ca4-0406-4612-82d5-7f07de81aeb0.log | jq
+    ```
+
+    Example output:
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: ditto.

File: gcs/COOPERATIVE_LOCKING.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+directory modification operations performed through `hadoop fs` command line
+interface on Cloud Storage.
+
+The key motivation behind this feature is a fact that directory mutations are
+not atomic on Cloud Storage, which means that competing directory mutations are
+susceptible to race conditions that can lead to data integrity issues (loss,
+duplication, etc).
+
+### How it works
+
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: Mention that the scope of lock is per bucket, dir modification operations in the same bucket will be sequential.

File: gcs/COOPERATIVE_LOCKING.md (hunk 8 above)
Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: Done.

File: gcs/COOPERATIVE_LOCKING.md (hunk 7 above)
Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: Done.

File: gcs/COOPERATIVE_LOCKING.md (hunk 9 above)
Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: Done.

File: gcs/COOPERATIVE_LOCKING.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+directory operations in `_lock/` folder:
+
+*   `all.lock` file
+
+    ```shell
+    gsutil ls -L gs://example-bucket/_lock/all.lock | \
+      grep "lock:" | awk '{print $2}' | base64 --decode | jq
+    ```
+
+    Example of the `all.lock` file `lock` metadata value:
```
Comment (functicons) [inline gcs/COOPERATIVE_LOCKING.md]: Explain the meaning of the samples, it will help users and developers understand how it works.

Comment (medb) [reply gcs/COOPERATIVE_LOCKING.md]: Done.

**Pull request merged**